### PR TITLE
Refactor achainable api using systemlabel

### LINKED ISSUE #1870: Create generalized assertions for new Achainable API
### Context
New Achainable API system label ref:
```
// POST /run/system-labels
export type PostRunSystemLabelRequest = {
  name: SystemLabelName;
  address: string;
  params: {
    [name: string]: any;
  };
  includeMetadata?: boolean;
};
```

### Task
As a follow up of this [PR](https://github.com/litentry/litentry-parachain/pull/1753), new assertions need to be added. Specifically, need to 
- find a more general architecture/structure to use Achainable API. 
- refactor current long list of Achainable interface

### Acceptance criteria
What's the definition of done, if applicable?

---
:heavy_check_mark: Please set appropriate **labels** and **assignees** if applicable.

### LINKED ISSUE #1913: A8 modification
### Context
Correct the implementation of A8

### Task
1. Change description into: "Gets the range of number of transactions a user has made for a specific token on all supported networks (invalid transactions are also counted)"

2. Define range of assertion results:
0, false
1-10, true
10-20, true
20-50, true
50-100, true
100-200, true
200-300, true
300-500, true
.. >500, true

### Acceptance criteria
correctly generates 1 false credential, 2 true credentials with different value of range

---
:heavy_check_mark: Please set appropriate **labels** and **assignees** if applicable.

### LINKED ISSUE #1914: A11,A10,A4,A7 modification
### Context
Correct implementation

### Task
1. change description into: "The length of time a user continues to hold a particular token (with particular threshold of token amount) "
2. define range of query period to a half year period:
2017/01/01 0:00 UTC - 2017/06/01 0:00 UTC
2017/06/01 0:00 UTC - 2017/12/31 0:00 UTC
2018/01/01 0:00 UTC - 2018/06/01 0:00 UTC
2018/06/01 0:00 UTC - 2018/12/31 0:00 UTC
2019/01/01 0:00 UTC - 2019/06/01 0:00 UTC
2019/06/01 0:00 UTC - 2019/12/31 0:00 UTC
2020/01/01 0:00 UTC - 2020/06/01 0:00 UTC
2020/06/01 0:00 UTC - 2020/12/31 0:00 UTC
2021/01/01 0:00 UTC - 2021/06/01 0:00 UTC
2022/06/01 0:00 UTC - 2022/12/31 0:00 UTC
2023/01/01 0:00 UTC - 2023/06/01 0:00 UTC
2023/06/01 0:00 UTC - now

### Acceptance criteria
True: returns the longest range of period
False: when user doesn't meet any of these periods

---
:heavy_check_mark: Please set appropriate **labels** and **assignees** if applicable.

### PR DESCRIPTION
**Pull request by outofxxx**
This is a PR with significant updates, focusing on three aspects:

0. Refactor achainbale api by using `SystemLabel`, see [issue-1870](https://github.com/litentry/litentry-parachain/issues/1870)

1. There have been changes to the query timeline on A4/7/10/11, see [issue-1914](https://github.com/litentry/litentry-parachain/issues/1914)

2. When processing 0 transaction volume, A8 returns false , also redefine range of assertion results. See [issue-1913](https://github.com/litentry/litentry-parachain/issues/1913)

### CODE REVIEW

File: primitives/core/src/network.rs [hunk 1]
```diff
@@ -78,6 +78,24 @@ impl Web3Network {
 	}
 }
 
+impl Web3Network {
+	/// Achainable server support lowercase ONLY.
```
Comment (kziemianek) [inline primitives/core/src/network.rs]: I would suggest putting `Achainable` related code into `Achainable` context, probably `achainable.rs`.

What about adding this function to `achainable.rs` ?

```
pub fn web3_network_to_chain(network: &Web3Network) -> &'static str {
	match network {
		Web3Network::Polkadot => "polkadot",
		Web3Network::Kusama => "kusama",
		Web3Network::Litentry => "litentry",
		Web3Network::Litmus => "litmus",
		Web3Network::LitentryRococo => "litentry_rococo",
		Web3Network::Khala => "khala",
		Web3Network::SubstrateTestnet => "substrate_testnet",
		Web3Network::Ethereum => "ethereum",
		Web3Network::Polygon => "polygon",
		Web3Network::BSC => "bsc",
	}
}
```

File: tee-worker/litentry/core/assertion-build/src/a10.rs [hunk 2]
```diff
@@ -53,13 +49,20 @@ pub fn build(req: &AssertionBuildRequest, min_balance: ParameterString) -> Resul
...
-	for index in 0..ASSERTION_FROM_DATE.len() {
+	for (index, date) in ASSERTION_FROM_DATE.iter().enumerate() {
 		if is_hold {
 			break
 		}
 
 		for address in &addresses {
-			match client.is_holder(&req.assertion, address, index) {
+			let holding = ParamsBasicTypeWithAmountHoding::new(
+				"ethereum".into(),
```
Comment (kziemianek) [inline tee-worker/litentry/core/assertion-build/src/a10.rs]: What do you think about using web3network here and mapping it into achainable chain ? With that we will reduce the amount of these consts spread across the code.

File: tee-worker/litentry/core/assertion-build/src/a11.rs [hunk 3]
```diff
@@ -49,13 +48,20 @@ pub fn build(req: &AssertionBuildRequest, min_balance: ParameterString) -> Resul
...
-	for index in 0..ASSERTION_FROM_DATE.len() {
+	for (index, date) in ASSERTION_FROM_DATE.iter().enumerate() {
 		if is_hold {
 			break
 		}
 
 		for address in &addresses {
-			match client.is_holder(&req.assertion, address, index) {
+			let holding = ParamsBasicTypeWithAmountHoding::new(
+				"ethereum".into(),
```
Comment (kziemianek) [inline tee-worker/litentry/core/assertion-build/src/a11.rs]: Same as above

File: tee-worker/litentry/core/assertion-build/src/a7.rs [hunk 4]
```diff
@@ -49,13 +48,20 @@ pub fn build(req: &AssertionBuildRequest, min_balance: ParameterString) -> Resul
...
-	for index in 0..ASSERTION_FROM_DATE.len() {
+	for (index, date) in ASSERTION_FROM_DATE.iter().enumerate() {
 		if is_hold {
 			break
 		}
 
 		for address in &addresses {
-			match client.is_holder(&req.assertion, address, index) {
+			let holding = ParamsBasicTypeWithAmountHoding::new(
+				"polkadot".into(),
```
Comment (kziemianek) [inline tee-worker/litentry/core/assertion-build/src/a7.rs]: same as above 🙃

File: primitives/core/src/assertion.rs [hunk 5]
```diff
@@ -74,12 +74,19 @@ impl Assertion {
 	}
 }
 
-pub const ASSERTION_FROM_DATE: [&str; 7] = [
+pub const ASSERTION_FROM_DATE: [&str; 14] = [
```
Comment (arieljbirnbaum) [inline primitives/core/src/assertion.rs:77]: Isn't that gonna be a _lot_ of requests? :sweat:

File: tee-worker/litentry/core/assertion-build/src/a4.rs
Comment (arieljbirnbaum) [inline tee-worker/litentry/core/assertion-build/src/a4.rs:1]: Please update the `TODO` comment (~ line 142):
```
// It is need to request TDF 2 (substrate+evm networks) * 7 (ASSERTION_FROM_DATE) = 14 http requests.
```
This part just got even worse :D _and_ it also needs to be multiplied by the number of addresses.

We could mitigate this by changing the loop to a binary search; let me know if you want to look into this. Alternatively, please link a followup issue in the comment :)

File: tee-worker/litentry/core/data-providers/src/achainable.rs [hunk 6]
```diff
@@ -155,196 +92,428 @@ pub trait AchainableResultParser {
...
+	pub chain: String,
+	pub amount: String,
+	pub date: String,
+
+	#[serde(skip_serializing_if = "Option::is_none")]
+	pub token: Option<String>,
+}
+
+impl ParamsBasicTypeWithAmountHoding {
+	pub fn new(netwok: &Web3Network, amount: String, date: String, token: Option<String>) -> Self {
```
Comment (arieljbirnbaum) [inline tee-worker/litentry/core/data-providers/src/achainable.rs]: :nail_care: just spelling:

```suggestion
impl ParamsBasicTypeWithAmountHolding {
	pub fn new(network: &Web3Network, amount: String, date: String, token: Option<String>) -> Self {
```

File: tee-worker/litentry/core/data-providers/src/achainable.rs [hunk 7]
```diff
@@ -155,196 +92,428 @@ pub trait AchainableResultParser {
...
+		let chain = web3_network_to_chain(netwok);
+		let name = if token.is_some() {
```
Comment (arieljbirnbaum) [inline tee-worker/litentry/core/data-providers/src/achainable.rs:212]: Just to be sure, the `chain` should _not_ be part of the name?

File: primitives/core/src/assertion.rs [hunk 8]
```diff
@@ -74,12 +74,19 @@ impl Assertion {
...
 	"2017-01-01",
+	"2017-06-01",
```
Comment (arieljbirnbaum) [inline primitives/core/src/assertion.rs]: Q: Is the midpoint supposed to be `06-01` and not `07-01`? :thinking:

File: tee-worker/litentry/core/assertion-build/src/a4.rs
Comment (arieljbirnbaum) [reply tee-worker/litentry/core/assertion-build/src/a4.rs:1]: Ah, there's an issue for this already: #1655 :D

File: primitives/core/src/assertion.rs (hunk 8 above)
Comment (BillyWooo) [reply primitives/core/src/assertion.rs]: good catch. Should be 07-01

File: tee-worker/litentry/core/data-providers/src/achainable.rs (hunk 7 above)
Comment (outofxxx) [reply tee-worker/litentry/core/data-providers/src/achainable.rs:212]: no, they are different fields

File: tee-worker/litentry/core/assertion-build/src/a4.rs
Comment (BillyWooo) [reply tee-worker/litentry/core/assertion-build/src/a4.rs:1]: Yeah, Achainable doesn't support vector of addresses any more. Look like we need to insist now.

File: primitives/core/src/assertion.rs (hunk 5 above)
Comment (BillyWooo) [reply primitives/core/src/assertion.rs:77]: No. Refer to our offline discussion.

**Review by BillyWooo — APPROVED** [review]
@grumpygreenguy is right. We still need to discuss about the `date` and `address` double loop topic.

**Review by BillyWooo — CHANGES_REQUESTED** [review]
Let's discuss about the `date` and `address` double loop before merge this PR.

**Review by BillyWooo — APPROVED** [review]

**Comment by BillyWooo** [comment]
> Let's discuss about the `date` and `address` double loop before merge this PR.

Let's ask our data provider first, before putting any extra effort to improve performance by ourselves.

**Pull request merged**